libp2p/transports/tcptransport.nim

```diff
@@ -149,31 +129,74 @@ method close*(t: TcpTransport) {.async, gcsafe.} =
     trace "transport stopped"
     inc getTcpTransportTracker().closed
   except CatchableError as exc:
-    trace "error shutting down tcp transport", exc = exc.msg
+    let useExc {.used.} = exc
+    trace "error shutting down tcp transport", errMsg = exc.msg
 
-method listen*(t: TcpTransport,
-               ma: MultiAddress,
-               handler: ConnHandler):
-               Future[Future[void]] {.async, gcsafe.} =
-  discard await procCall Transport(t).listen(ma, handler) # call base
+method listen*(t: TcpTransport, ma: MultiAddress,
+               handler: ConnHandler) {.async, gcsafe.} =
+  discard procCall Transport(t).listen(ma, handler) # call base
 
   ## listen on the transport
-  t.server = createStreamServer(t.ma, connCb, t.flags, t)
-  t.server.start()
+  t.server = createStreamServer(t.ma, t.flags, t)
 
   # always get the resolved address in case we're bound to 0.0.0.0:0
   t.ma = MultiAddress.init(t.server.sock.getLocalAddress()).tryGet()
-  result = t.server.join()
-  trace "started node on", address = t.ma
+  trace "Listen started on", address = t.ma
+
+  while true:
+    try:
+      let transp = await t.server.accept()
+      trace "Received incoming connection", address = $transp.remoteAddress
+      # we don't need result connection in this case as it's added inside
+      # connHandler
+      try:
+        discard t.connHandler(transp, false)
+      except CancelledError as exc:
+        raise exc
+      except CatchableError as exc:
+        let useExc {.used.} = exc
+        debug "listen: connection setup failed", errMsg = exc.msg
+        if not transp.closed:
+          await transp.closeWait()
```

Comment:
@dryajov i have copied your code from here
https://github.com/status-im/nim-libp2p/pull/247/files#diff-54deac2904add92d29f284d9eb946cc7L98-L108

The only change is `close` was replaced with `closeWait()`. And because `transport.close()` operation is not raising anymore i have removed `try/catch` block.

Reply:
ok, I confused this with the `server.close()`, this one if fine then.